test/demo/utils/RemoteDirectory.ts

```diff
@@ -0,0 +1,118 @@
+import {Identity} from 'openfin/_v2/main';
+
+import {Application, RemoteSnippetsDirectoryCollection, StoredApplicationsDirectoryCollection} from '../../../src/client/main';
+
+import {ofBrowser, handlePuppeteerError} from './fdc3RemoteExecution';
+import {TestWindowContext} from './ofPuppeteer';
+
+export class RemoteDirectory {
+    private readonly _executionTarget: Identity;
+    private readonly _index: number;
+
+    public readonly remoteSnippets: RemoteRemoteSnippetsDirectoryCollection;
+    public readonly storedApplications: RemoteStoredApplicationDirectoryCollection;
+
+    public constructor(executionTarget: Identity, index: number) {
+        this._executionTarget = executionTarget;
+        this._index = index;
+
+        this.remoteSnippets = new RemoteRemoteSnippetsDirectoryCollection(executionTarget, this._index);
+        this.storedApplications = new RemoteStoredApplicationDirectoryCollection(executionTarget, this._index);
+    }
+
+    public get sourceVersion(): Promise<number> {
+        return ofBrowser.executeOnWindow(this._executionTarget, function (this: TestWindowContext, remoteIndex: number): number {
+            return this.directories[remoteIndex].sourceVersion;
+        }, this._index).catch(handlePuppeteerError);
+    }
+}
+
+class RemoteDirectoryCollection<
+    T extends (RemoteSnippetsDirectoryCollection & {type: 'remoteSnippets'}) | (StoredApplicationsDirectoryCollection & {type: 'storedApplications'})
+> {
+    protected readonly _executionTarget: Identity;
+    protected readonly _index: number;
+
+    private readonly _aspect: T['type'];
+
+    public constructor(executionTarget: Identity, index: number, aspect: T['type']) {
+        this._executionTarget = executionTarget;
+        this._index = index;
+        this._aspect = aspect;
+    }
+
+    public get source(): Promise<T['source']> {
+        return ofBrowser.executeOnWindow(this._executionTarget, function (
+            this: TestWindowContext,
+            remoteIndex: number,
+            remoteAspect: T['type']
+        ): T['source'] {
+            return this.directories[remoteIndex][remoteAspect].source as T['source'];
+        }, this._index, this._aspect).catch(handlePuppeteerError);
+    }
+
+    public async add(arg: Parameters<T['add']>[0]): Promise<void> {
+        return ofBrowser.executeOnWindow(this._executionTarget, function (
+            this: TestWindowContext,
+            remoteIndex: number,
+            remoteAspect: T['type'],
+            remoteArg: Parameters<T['add']>[0]
+        ): void {
+            this.directories[remoteIndex][remoteAspect].add(remoteArg as any);
```

Comment:
Open to ideas on how to get rid of these `any`s

Reply:
I think we're getting beyond the limits of Typescript here. May as well leave as is.